### DISCUSSION TITLE
Do people use it in production?

### DISCUSSION THREAD

**Discussion by buehler** [dotnet/dotnet-operator-sdk#215] [Q&A] [answered]
I'd like to ask if this SDK (and coded operators) are used in production or near production environments.

I used an older version (v2) of the SDK to code an operator and sometimes I experienced memory errors. I assume they have been eliminated with the move to instantiated scopes and calls to controllers.

But I want to know if this project has any use (other than academic).

[the remark this answers is not in the document]
**Reply by tomkerkhove** [reply]
Not yet, but I'm planning on porting promitor.io over.

But time is a pain, as always.

[the remark this answers is not in the document]
**Reply by buehler** [reply]
we all know that problem 😏

[the remark this answers is not in the document]
**Reply by erin-allison** [reply]
For now, the answer is sadly no. I have a small number of private projects which may or may not come to fruition, and some of those probably will.

[the remark this answers is not in the document]
**Reply by ocdi** [reply]
@buehler I am using it in production, however I haven't had a change to update it to newer releases since my brief involvement a while back. I do need to update to use the health checks, occasionally the watch dies and doesn't resume. The fix is just restart the operator pod and it all kicks back into life.

[the remark this answers is not in the document]
**Reply by erin-allison** [reply]
> 
> 
> @buehler I am using it in production, however I haven't had a change to update it to newer releases since my brief involvement a while back. I do need to update to use the health checks, occasionally the watch dies and doesn't resume. The fix is just restart the operator pod and it all kicks back into life.

I'd be interested in what your logs say about that - I have had some running through the API Proxy that Rancher uses so watches get cut off after a short period (ingress-nginx has a low value for max time by default), but the library definitely has reconnect logic that I can see logged if I have my verbosity up.

[the remark this answers is not in the document]
**Reply by erin-allison** [reply]
Also I guess if it's a really old release, the support for that could be iffy, either in https://github.com/buehler/dotnet-kubernetes-client or (more likely) in Google's auto-generated client that buehler's builds on top of.

Sounds like an interesting issue to try and hunt down if it is present still on a new release.

[the remark this answers is not in the document]
**Reply by buehler** [reply]
Definitely! I experienced some issues - like the described - as well. Sometimes, the operator died on its own. I'm going to update my operator that we used in production to see if this is still an issue.

[the remark this answers is not in the document]
**Reply by buehler** [reply]
But all in all, nice that it is being used :-)

[the remark this answers is not in the document]
**Reply by alexmg** [reply]
I have been using this in a private project but have struggled to find time to push it forward. I cannot comment regarding reliability because I have not used the SDK in a production like setting yet.

It is great to have something like this for the .NET folks that removes the need to write so much plumbing code. It has helped me get my experiment much further along in the limited time I have had to work on it.

Seeing this discussion has encouraged me to go back to my experiment and push it that little bit further along. 😄

[the remark this answers is not in the document]
**Reply by ghost** [reply]
Not used in production yet, but we plan to very soon.

[the remark this answers is not in the document]
**Reply by buehler** [reply]
I just improved the developer experience when using validators / mutators :-)
you can now add a "localtunnel" connection directly in the config and while debugging, the operator registers itself in kubernetes with the localtunnel connection as the target for webhooks. now you can just run the operator locally and forget about ngrok or whatever proxy you used.

[the remark this answers is not in the document]
**Reply by ewassef** [reply]
we use this on 3 different operators in production. One of the operators is what we call a "user cluster operator" so it's deployed o 100s of clusters. Great work!

**Comment by McSoda** [comment]
Hi @all,
I know this question is quite old, but I hope my answer will add value.
I just developed an operator with this sdk and will start to use it in production.
The operator is quite simple: it measure waiting messages in a RabbitMQ queue and scale a deployment to consume them.
If anyone interressed, I can share code.

I struggled with some problems (essentially with a date field) not all solved as I would hope but the operator is working and I will use it in this state and try to enhance functionnalities.

Kind regards,

**Comment by virgilp** [comment]
I plan to use it in production but I ran into a few problems with the deployment & security aspect of it. I opened https://github.com/buehler/dotnet-operator-sdk/issues/567 as well as a PR for that issue.... Maybe it's the most elegant solution, works for me I guess and I still think it's better than  downloading stuff at runtime.... but indeed it's less convenient for development. However, it's questionable whether development convenience should take precedence over production security/deployment versatility.
